### PR TITLE
Refactor image pulling to use goroutines for concurrency

### PR DESCRIPTION
**Pull request by corang**
## Description

Concurrently retrieves image metadata and layers during `package create`.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [x] Test, docs, adr added or updated as needed
- [x] [Contributor Guide Steps](https://github.com/defenseunicorns/zarf/blob/main/CONTRIBUTING.md#developer-workflow) followed

### CODE REVIEW

**Comment by corang** [comment]
bigbang example package generation
modded run times:
```
real    2m35.223s
user    6m15.280s
sys     0m59.717s
---
real    2m36.357s
user    6m20.161s
sys     0m59.530s
---
real    2m35.577s
user    6m24.044s
sys     0m56.640s
```
0.28.0 run times:
```
real    3m35.299s
user    9m21.381s
sys     0m47.760s
---
real    3m34.414s
user    9m33.166s
sys     0m46.836s
---
real    3m41.898s
user    8m47.642s
sys     0m44.218s
```
0.28.0 skip sbom run times:
```
real    2m28.487s
user    2m38.983s
sys     0m12.519s
---
real    2m22.387s
user    2m37.232s
sys     0m13.272s
---
real    2m22.329s
user    2m36.669s
sys     0m12.810s
```

**Comment by corang** [comment]
tldr:
- goroutine    avg 2m35.719s
- v28 normal avg 3m37.204s
- v28 skip      avg 2m24.401s
sbom creation time is now 11s vs 1m13s on my machine with the bigbang example package

File: src/internal/packager/sbom/catalog.go [hunk 1]
```diff
@@ -70,30 +70,59 @@ func Catalog(componentSBOMs map[string]*types.ComponentSBOM, imgList []string, t
...
+		select {
+		// If there was an error generating an SBOM, write the error to the spinner
+		case err := <-imageError:
+			builder.spinner.Errorf(err.err, err.message)
+			return err.err
+		// If there is a string in the imageProgress channel we know that an SBOM 
+		// was generated for an image and we can update the spinner
+		case tag := <-imageProgress:
+			builder.spinner.Updatef("Creating image SBOMs (%d of %d): %s", i, len(imgList), tag)
+		}
```
Comment (Racer159) [inline src/internal/packager/sbom/catalog.go:125]: Can we do this for components as well?

**Review by Racer159 — CHANGES_REQUESTED** [review]

File: src/internal/packager/sbom/catalog.go [hunk 2]
```diff
@@ -70,30 +70,52 @@ func Catalog(componentSBOMs map[string]*types.ComponentSBOM, imgList []string, t
...
+		}(tag)
+	}
 
-		currImage++
+	// Wait for all images to be done generating SBOMs
+	for i := 0; i < len(imgList); i++ {
+		select {
+		// If there was an error generating an SBOM, write the error to the spinner
+		case err := <-imageError:
+			builder.spinner.Errorf(err, "Unable to load the image to generate an SBOM")
```
[the remark this answers is not in the document]
Comment (Racer159) [reply src/internal/packager/sbom/catalog.go]: This was done before the review finished

File: src/internal/packager/sbom/catalog.go [hunk 3]
```diff
@@ -70,30 +70,59 @@ func Catalog(componentSBOMs map[string]*types.ComponentSBOM, imgList []string, t
...
 
-		jsonData, err := builder.createImageSBOM(img, tag)
-		if err != nil {
-			builder.spinner.Errorf(err, "Unable to create SBOM for image %s", tag)
-			return err
-		}
+	type errorWithMessage struct {
+		err     error
+		message string
+	}
```
Comment (Racer159) [inline src/internal/packager/sbom/catalog.go:81]: this might also be useful in message.go as a more common pattern

**Review by Racer159 — CHANGES_REQUESTED** [review]

**Comment by corang** [comment]
hijacked the PR to include image pull concurrency too @Racer159 Can move it to another PR if it's not wanted here.

File: src/internal/packager/sbom/catalog.go (hunk 3 above)
Comment (corang) [reply src/internal/packager/sbom/catalog.go]: made concurrency stuff reusable and put that struct in message.go

File: src/internal/packager/sbom/catalog.go (hunk 1 above)
Comment (corang) [reply src/internal/packager/sbom/catalog.go:134]: for image pulling too!

**Comment by corang** [comment]
Don't know why ci is failing, same commands work fine on my local machine

**Review by Racer159 — CHANGES_REQUESTED** [review]

File: src/pkg/utils/concurrency.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+		case err := <-ct.ErrorChan:
+			ct.Cancel()
+			errResult := onError(err)
+			return errResult
+		case  <-ct.ProgressChan:
+		}
+	}
+	ct.waitGroup.Wait()
+	return nil
+}
```
Comment (Noxsios) [inline src/pkg/utils/concurrency.go:100]: isnt this technically a  `return ct.WaitWithProgress(nil, onError)`?

**Review by Noxsios — CHANGES_REQUESTED** [review]

File: src/internal/packager/images/pull.go [hunk 5]
```diff
@@ -99,52 +150,291 @@ func (i *ImgConfig) PullAll() error {
...
+
+			readCloser, err := layer.Compressed()
+			if err != nil {
+				layerWritingConcurrency.ErrorChan <- err
+				return
+			}
+
+			// Create the directory for the blob if it doesn't exist
+			dir := filepath.Join(string(cranePath), "blobs", digest.Algorithm)
+			if err := os.MkdirAll(dir, os.ModePerm); err != nil && !os.IsExist(err) {
```
Comment (Racer159) [inline src/internal/packager/images/pull.go:251]:
```suggestion
			if err := utils.CreateDirectory(dir, os.ModePerm); err != nil {
```
Per @Noxsios 's comment (he may have selected the wrong line)

**Review by Racer159 — CHANGES_REQUESTED** [review]

File: src/pkg/utils/concurrency.go (hunk 4 above)
Comment (corang) [reply src/pkg/utils/concurrency.go:100]: nil pointer dereference if you do that

File: src/internal/packager/images/pull.go (hunk 5 above)
Comment (corang) [reply src/internal/packager/images/pull.go]: changed

File: src/pkg/utils/concurrency.go (hunk 4 above)
Comment (corang) [reply src/pkg/utils/concurrency.go:100]: Its things like this that make me wish go had overloading for functions

**Review by Noxsios — APPROVED** [review]

**Review by Racer159 — APPROVED** [review]
